src/lib/pubkey/hss_lms/hss.cpp

```diff
@@ -0,0 +1,418 @@
+      }
+   }
+
+   return sig;
+}
+
+LMS_PrivateKey HSS_LMS_PrivateKeyInternal::hss_derive_root_lms_private_key() const {
+   auto& top_params = hss_params().params_at_level(HSS_Level(0));
+   return LMS_PrivateKey(top_params.lms_params(), top_params.lmots_params(), m_identifier, m_hss_seed);
+}
+
+LMS_PrivateKey HSS_LMS_PrivateKeyInternal::hss_derive_child_lms_private_key(
+   const HSS_LMS_Params::LMS_LMOTS_Params_Pair& child_lms_lmots_params,
+   const LMS_PrivateKey& parent_sk,
+   LMS_Tree_Node_Idx parent_q) {
+   const auto hash = HashFunction::create_or_throw(child_lms_lmots_params.lms_params().hash_name());
+
+   // CHILD_SEED = H( PARENT_I || PARENT_Q || SEED_CHILD_SEED || 0xff || PARENT_SEED )
+   PseudorandomKeyGeneration seed_generator(parent_sk.identifier());
+   seed_generator.set_q(parent_q.get());
+   seed_generator.set_i(SEED_CHILD_SEED);
+   seed_generator.set_j(0xff);
+   auto child_seed = seed_generator.gen<LMS_Seed>(*hash, parent_sk.seed());
+
+   // CHILD_I = H( PARENT_I || PARENT_Q || SEED_CHILD_I || 0xff || PARENT_SEED )
+   seed_generator.set_i(SEED_CHILD_I);
+   auto child_identifier = seed_generator.gen<LMS_Identifier>(*hash, parent_sk.seed());
+   child_identifier.resize(LMS_IDENTIFIER_LEN);
+
+   return LMS_PrivateKey(child_lms_lmots_params.lms_params(),
+                         child_lms_lmots_params.lmots_params(),
+                         std::move(child_identifier),
+                         std::move(child_seed));
+}
+
+HSS_LMS_PublicKeyInternal HSS_LMS_PublicKeyInternal::create(const HSS_LMS_PrivateKeyInternal& hss_sk) {
+   auto& hss_params = hss_sk.hss_params();
+
+   const auto root_sk = hss_sk.hss_derive_root_lms_private_key();
+   LMS_PublicKey top_pub_key = LMS_PublicKey(root_sk);
+
+   return HSS_LMS_PublicKeyInternal(hss_params.L(), std::move(top_pub_key));
+}
+
+std::shared_ptr<HSS_LMS_PublicKeyInternal> HSS_LMS_PublicKeyInternal::from_bytes_or_throw(
+   std::span<const uint8_t> key_bytes) {
+   if(key_bytes.size() < sizeof(HSS_Level)) {
+      throw Decoding_Error("Too few public key bytes.");
+   }
+   BufferSlicer slicer(key_bytes);
+
+   const auto L = load_be<HSS_Level>(slicer.take<sizeof(HSS_Level)>());
+   if(L > HSS_MAX_LEVELS) {
+      throw Decoding_Error("Invalid number of HSS layers in public HSS-LMS key.");
+   }
+
+   LMS_PublicKey lms_pub_key = LMS_PublicKey::from_bytes_or_throw(slicer);
+
+   if(!slicer.empty()) {
+      throw Decoding_Error("Public HSS-LMS key contains more bytes than expected.");
+   }
+   return std::make_shared<HSS_LMS_PublicKeyInternal>(L, std::move(lms_pub_key));
+}
+
+std::vector<uint8_t> HSS_LMS_PublicKeyInternal::to_bytes() const {
+   std::vector<uint8_t> lms_pub_key_bytes = m_top_lms_pub_key.to_bytes();
+   std::vector<uint8_t> bytes(sizeof(m_L) + lms_pub_key_bytes.size());
+   BufferStuffer stuffer(bytes);
+
+   stuffer.append(store_be(m_L));
+   stuffer.append(lms_pub_key_bytes);
+   BOTAN_ASSERT_NOMSG(stuffer.full());
+
+   return bytes;
```

Comment:
```suggestion
   return concat<std::vector<uint8_t>>(store_be(m_L), m_top_lms_pub_key.to_bytes());
```

... this should work now (using https://github.com/randombit/botan/pull/3994).